spring-integration-core/src/main/java/org/springframework/integration/support/json/EmbeddedJsonHeadersMessageMapper.java

```diff
@@ -0,0 +1,218 @@
+ * as JSON. For messages with {@code byte[]} payloads, if rendered as JSON, Jackson
+ * performs Base64 conversion on the bytes. If the {@link #setRawBytes(boolean) rawBytes}
+ * property is true (default), the result has the form
+ * &lt;headersLen&gt;&lt;headers&gt;&lt;payloadLen&gt;&lt;payload&gt;; with the headers
+ * rendered in JSON and the payload unchanged.
+ * <p>
+ * By default, all headers are included; you can provide regular expressions to specify a
+ * subset of headers.
+ * <p>
+ * If neither expected format is detected, or an error occurs during conversion, the
+ * payload of the message is the original {@code byte[]}.
+ * <p>
+ * <b>IMPORTANT</b>
+ * <p>
+ * The default object mapper will only deserialize classes in certain packages.
+ *
+ * <pre class=code>
+ *	"java.util",
+ *	"java.lang",
+ *	"org.springframework.messaging.support",
+ *	"org.springframework.integration.support",
+ *	"org.springframework.integration.message",
+ *	"org.springframework.integration.store"
+ * </pre>
+ * <p>
+ * To add more packages, create an object mapper using
+ * {@link JacksonJsonUtils#messagingAwareMapper(String...)}.
+ * <p>
+ * A constructor is provided allowing the provision of such a configured object mapper.
+ *
+ * @author Gary Russell
+ * @since 5.0
+ *
+ */
+public class EmbeddedJsonHeadersMessageMapper implements BytesMessageMapper {
+
+	private final ObjectMapper objectMapper;
+
+	private final List<Pattern> headerPatterns;
+
+	private final boolean allHeaders;
+
+	private boolean rawBytes = true;
+
+	/**
+	 * Construct an instance that embeds all headers, using the default
+	 * JSON Object mapper.
+	 */
+	public EmbeddedJsonHeadersMessageMapper() {
+		this(Pattern.compile(".*"));
+	}
+
+	/**
+	 * Construct an instance that embeds headers matching the supplied patterns, using
+	 * the default JSON object mapper.
+	 * @param headerPatterns the patterns.
+	 */
+	public EmbeddedJsonHeadersMessageMapper(Pattern... headerPatterns) {
+		this(JacksonJsonUtils.messagingAwareMapper(), headerPatterns);
+	}
+
+	/**
+	 * Construct an instance that embeds all headers, using the
+	 * supplied JSON object mapper.
+	 * @param objectMapper the object mapper.
+	 */
+	public EmbeddedJsonHeadersMessageMapper(ObjectMapper objectMapper) {
+		this(objectMapper, Pattern.compile(".*"));
+	}
+
+	/**
+	 * Construct an instance that embeds headers matching the supplied patterns using the
+	 * supplied JSON object mapper.
+	 * @param objectMapper the object mapper.
+	 * @param headerPatterns the patterns.
+	 */
+	public EmbeddedJsonHeadersMessageMapper(ObjectMapper objectMapper, Pattern... headerPatterns) {
+		this.objectMapper = objectMapper;
+		this.headerPatterns = Arrays.asList(headerPatterns);
+		this.allHeaders = this.headerPatterns.size() == 1 && this.headerPatterns.get(0).pattern().equals(".*");
+	}
+
+	/**
+	 * Set to false to encode byte[] payloads as base64 in JSON. When true,
+	 * byte[] payloads are rendered as discussed in the class description.
```

Comment:
IMO that's not JavaDoc language. I'm fine to have the copy of the sentence from the class level here.